tools/syz-trace2syz/proggen/vma_call_handlers.go

```diff
@@ -0,0 +1,350 @@
+// Copyright 2018 syzkaller project authors. All rights reserved.
+// Use of this source code is governed by Apache 2 LICENSE that can be found in the LICENSE file.
+
+package proggen
+
+import (
+	"github.com/google/syzkaller/pkg/log"
+	"github.com/google/syzkaller/prog"
+	"github.com/google/syzkaller/tools/syz-trace2syz/parser"
+)
+
+func parseMemoryCall(ctx *Context) *prog.Call {
+	syzCall := ctx.CurrentSyzCall
+	straceCall := ctx.CurrentStraceCall
+
+	switch straceCall.CallName {
+	case "mmap":
+		return parseMmap(syzCall.Meta, straceCall, ctx)
+	case "mremap":
+		return parseMremap(syzCall.Meta, straceCall, ctx)
+	case "msync":
+		return parseMsync(syzCall.Meta, straceCall, ctx)
+	case "mprotect":
+		return parseMprotect(syzCall.Meta, straceCall, ctx)
+	case "munmap":
+		return parseMunmap(syzCall.Meta, straceCall, ctx)
+	case "madvise":
+		return parseMadvise(syzCall.Meta, straceCall, ctx)
+	case "mlock":
+		return parseMlock(syzCall.Meta, straceCall, ctx)
+	case "munlock":
+		return parseMunlock(syzCall.Meta, straceCall, ctx)
+	case "shmat":
+		return parseShmat(syzCall.Meta, straceCall, ctx)
+	}
+	return nil
+}
+
+func parseMmap(mmap *prog.Syscall, syscall *parser.Syscall, ctx *Context) *prog.Call {
+	call := &prog.Call{
+		Meta: mmap,
+		Ret:  prog.MakeReturnArg(mmap.Ret),
+	}
+	ctx.CurrentSyzCall = call
+	pageSize := ctx.Target.PageSize
+
+	length := (parseLength(syscall.Args[1], ctx) + pageSize - 1) / pageSize * pageSize // RoundUp PageSize
+	log.Logf(3, "mmap call: %#v requires %d memory", syscall, length)
+	addrArg, start := parseAddr(length, mmap.Args[0], syscall.Args[0], ctx)
+	lengthArg := prog.MakeConstArg(mmap.Args[1], length)
+	protArg := parseFlags(mmap.Args[2], syscall.Args[2], ctx, false)
+	flagArg := parseFlags(mmap.Args[3], syscall.Args[3], ctx, true)
+	fdArg := parseFd(mmap.Args[4], syscall.Args[4], ctx)
+
+	call.Args = []prog.Arg{
+		addrArg,
+		lengthArg,
+		protArg,
+		flagArg,
+		fdArg,
+		prog.MakeConstArg(mmap.Args[5], 0),
+	}
+	ctx.Tracker.createMapping(call, len(ctx.Prog.Calls), call.Args[0], start, start+length)
+	return call
+}
+
+func parseMremap(mremap *prog.Syscall, syscall *parser.Syscall, ctx *Context) *prog.Call {
+	call := &prog.Call{
+		Meta: mremap,
+		Ret:  prog.MakeReturnArg(mremap.Ret),
+	}
+	ctx.CurrentSyzCall = call
+
+	oldAddrArg, start := parseAddr(ctx.Target.PageSize, mremap.Args[0], syscall.Args[0], ctx)
+	oldSz := parseLength(syscall.Args[1], ctx)
+	newSz := parseLength(syscall.Args[2], ctx)
+	oldSzArg := prog.MakeConstArg(mremap.Args[1], oldSz)
```

Comment:
No this worked because we fill out memory after generating the program. I think doing a post hook is probably better than having a separate file for these calls. I'll add that to my TODOs :)

Reply:
This is called from parseMemoryCall, which is called from genCall, which is called from GenSyzProg, which does AssignSizesCall afterwards. AssignSizesCall should overwrite all len arguments. What am I missing?

Reply:
Sorry AssignSizesCall does fill out the length argument but when we call MakeVmaPointerArg we provide it the correct size. See [here](https://github.com/shankarapailoor/syzkaller/blob/master/tools/syz-trace2syz/proggen/vma_call_handlers.go#L283).

Reply:
We pass in ctx.Target.PageSize in most cases.